routes/runtime/github.rb

```diff
@@ -108,8 +108,28 @@ class Clover
           rescue Sequel::ValidationFailed, Sequel::UniqueConstraintViolation
             fail CloverError.new(409, "AlreadyExists", "A cache entry for #{scope} scope already exists with #{key} key and #{version} version.")
           end
-          upload_id = repository.blob_storage_client.create_multipart_upload(bucket: repository.bucket_name, key: entry.blob_key).upload_id
-          entry.update(upload_id: upload_id)
+
+          # Token creation on Cloudflare R2 takes time to propagate. Since that point is the
+          # first time we use the credentials, we are waiting it to be propagated. Note that,
+          # credential propagation will happen only while the bucket and token are being created
+          # initially. So, the retry block expected to run only while saving the first cache
+          # entry for a repository.
+          retries = 0
+          begin
+            upload_id = repository.blob_storage_client.create_multipart_upload(bucket: repository.bucket_name, key: entry.blob_key).upload_id
+            entry.update(upload_id: upload_id)
+          rescue Aws::S3::Errors::Unauthorized => ex
+            retries += 1
+            if retries < 3
+              # :nocov:
+              sleep(Config.test? ? 0 : 1)
+              # :nocov:
```

Comment:
Instead of branching here, we can mock the sleep in the tests like this:

```ruby
expect_any_instance_of(described_class).to receive(:sleep).with(1).twice
```

Reply:
We talked about it privately; since we freeze classes, we can't mock `sleep`.